helix-vcs/src/jj.rs

```diff
@@ -0,0 +1,282 @@
+        .arg("--repository")
+        .arg(repo)
+        .args([
+            "--ignore-working-copy",
+            "log",
+            "--color",
+            "never",
+            "--revisions",
+            "@", // Only display the current revision
+            "--no-graph",
+            "--no-pager",
+            "--template",
+            template,
+        ])
+        .output()?;
+
+    if !out.status.success() {
+        anyhow::bail!("jj log command executed but failed");
+    }
+
+    let out = String::from_utf8(out.stdout)?;
+
+    let rev = out
+        .lines()
+        .next()
+        // Contrary to git, if a JJ repo exists, it always has at least two revisions:
+        // the root (zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz), which cannot be focused, and the current
+        // one, which exists even for brand new repos.
+        .context("should always find at least one line")?;
+
+    Ok(Arc::new(ArcSwap::from_pointee(rev.into())))
+}
+
+pub(crate) fn for_each_changed_file(
+    repo: &Path,
+    callback: impl Fn(Result<FileChange>) -> bool,
+) -> Result<()> {
+    let out = Command::new("jj")
+        .arg("--repository")
+        .arg(repo)
+        .args([
+            "--ignore-working-copy",
+            "diff",
+            "--color",
+            "never",
+            "--revision",
+            "@", // Only display the current revision
+            "--no-pager",
+            "--types",
+        ])
+        .output()?;
+
+    if !out.status.success() {
+        anyhow::bail!("jj log command executed but failed");
+    }
+
+    let out = String::from_utf8(out.stdout)?;
+
+    for line in out.lines() {
+        let Some((status, path)) = line.split_once(' ') else {
+            continue;
+        };
+
+        let Some(change) = status_to_change(status, path) else {
+            continue;
+        };
+
+        if !callback(Ok(change)) {
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) fn open_repo(repo_path: &Path) -> Result<()> {
+    assert!(
+        repo_path.join(".jj").exists(),
+        "no .jj where one was expected: {}",
+        repo_path.display(),
+    );
+
+    let status = Command::new("jj")
+        .args(["--ignore-working-copy", "root"])
+        .stdout(std::process::Stdio::null())
+        .stderr(std::process::Stdio::null())
+        .status()?;
+
+    if status.success() {
+        Ok(())
+    } else {
+        anyhow::bail!("not a valid JJ repo")
+    }
+}
+
+/// Associate a status to a `FileChange`.
+fn status_to_change(status: &str, path: &str) -> Option<FileChange> {
+    if let rename @ Some(_) = find_rename(path) {
+        return rename;
+    }
+
+    // Syntax: <https://github.com/martinvonz/jj/blob/f9cfa5c9ce0eacd38e961c954e461e5e73067d22/cli/src/diff_util.rs#L97-L101>
+    Some(match status {
+        "FF" | "LL" | "CF" | "CL" | "FL" | "LF" => FileChange::Modified { path: path.into() },
+        "-F" | "-L" => FileChange::Untracked { path: path.into() },
```

Comment:
`Untracked` feels a bit off, would it make sense to add `Added` (not only for `jj`, but also in general)?

Reply:
For JJ specifically, I agree `Untracked` makes little sense, but it's better to reuse a variant that is highlighted as an addition than to add JJ-specific behaviour IMO, since JJ is very much the small player in the VCS space, I'm trying to avoid imposing too much maintenance burden on helix maintainers :)

Reply:
`jj` has a config option to tune the `autosnapshot`, so it _can_ have untracked files.

Reply:
Fair points! Indeed maybe just getting `jj` supported would be a good first step :)